File: src/type/infer.js
import { LambError } from "../syntax/lexer.js"
import { arrow, formatType, generalize, instantiate, list, option, primitive, prune, resetTypes, tuple, unify, variable } from "./system.js"

const int = primitive("int")
const float = primitive("float")
const bool = primitive("bool")
const string = primitive("string")
const char = primitive("char")
const unit = primitive("unit")

function constructors(level) {
  const optional = variable(level)
  return new Map([
    ["None", { variables: [optional], type: option(optional) }],
    ["Some", { variables: [optional], type: arrow(optional, option(optional)) }]
  ])
}

function bindPattern(pattern, expected, environment, level, names = new Set()) {
  pattern.type = prune(expected)
  if (pattern.kind === "pwildcard") return
  if (pattern.kind === "pvariable") {
    if (names.has(pattern.name)) throw new LambError(`variable ${pattern.name} is bound several times in this pattern`, pattern.start, pattern.end, "E_PATTERN_BINDING")
    names.add(pattern.name)
    environment.set(pattern.name, { variables: [], type: expected })
    return
  }
  if (pattern.kind === "pinteger") return unify(expected, int, pattern)
  if (pattern.kind === "pboolean") return unify(expected, bool, pattern)
  if (pattern.kind === "pstring") return unify(expected, string, pattern)
  if (pattern.kind === "pcharacter") return unify(expected, char, pattern)
  if (pattern.kind === "punit") return unify(expected, unit, pattern)
  if (pattern.kind === "pnil") return unify(expected, list(variable(level)), pattern)
  if (pattern.kind === "pcons") {
    const item = variable(level)
    unify(expected, list(item), pattern)
    bindPattern(pattern.head, item, environment, level, names)
    bindPattern(pattern.tail, list(item), environment, level, names)
    return
  }
  if (pattern.kind === "ptuple") {
    const items = pattern.items.map(() => variable(level))
    unify(expected, tuple(items), pattern)
    pattern.items.forEach((entry, index) => bindPattern(entry, items[index], environment, level, names))
    return
  }
  if (pattern.kind === "pconstructor") {
    if (pattern.name === "None") {
      if (pattern.argument) throw new LambError("constructor None expects no argument", pattern.start, pattern.end, "E_CONSTRUCTOR_ARITY")
      return unify(expected, option(variable(level)), pattern)
    }
    if (pattern.name === "Some") {
      if (!pattern.argument) throw new LambError("constructor Some expects one argument", pattern.start, pattern.end, "E_CONSTRUCTOR_ARITY")
      const item = variable(level)
      unify(expected, option(item), pattern)
      bindPattern(pattern.argument, item, environment, level, names)
      return
    }
    throw new LambError(`unknown constructor ${pattern.name}`, pattern.start, pattern.end, "E_CONSTRUCTOR")
  }
}

function inferNode(node, environment, level, constructorEnvironment) {
  let type
  if (node.kind === "integer") type = int
  else if (node.kind === "float") type = float
  else if (node.kind === "boolean") type = bool
  else if (node.kind === "string") type = string
  else if (node.kind === "character") type = char
  else if (node.kind === "unit") type = unit
  else if (node.kind === "variable") {
    const scheme = environment.get(node.name)
    if (!scheme) throw new LambError(`unbound value ${node.name}`, node.start, node.end, "E_UNBOUND_VALUE")
    type = instantiate(scheme, level)
  } else if (node.kind === "constructor") {
    const scheme = constructorEnvironment.get(node.name)
    if (!scheme) throw new LambError(`unknown constructor ${node.name}`, node.start, node.end, "E_CONSTRUCTOR")
    type = instantiate(scheme, level)
  } else if (node.kind === "unary") {
    const value = inferNode(node.value, environment, level, constructorEnvironment)
    unify(value, int, node.value)
    type = int
  } else if (node.kind === "binary") {
    const left = inferNode(node.left, environment, level, constructorEnvironment)
    const right = inferNode(node.right, environment, level, constructorEnvironment)
    if (["+", "-", "*", "/", "mod"].includes(node.operator)) {
      unify(left, int, node.left)
      unify(right, int, node.right)
      type = int
    } else if (node.operator === "^") {
      unify(left, string, node.left)
      unify(right, string, node.right)
      type = string
    } else if (["<", ">", "<=", ">="].includes(node.operator)) {
      unify(left, right, node)
      const resolved = prune(left)
      if (resolved.tag === "primitive" && !["int", "float", "string", "char"].includes(resolved.name)) throw new LambError(`operator ${node.operator} does not order ${formatType(resolved)}`, node.start, node.end, "E_OPERATOR")
      type = bool
    } else if (["&&", "||"].includes(node.operator)) {
      unify(left, bool, node.left)
      unify(right, bool, node.right)
      type = bool
    } else {
      unify(left, right, node)
      type = bool
    }
  } else if (node.kind === "cons") {
    const head = inferNode(node.head, environment, level, constructorEnvironment)
    const tail = inferNode(node.tail, environment, level, constructorEnvironment)
    unify(tail, list(head), node.tail)
    type = tail
  } else if (node.kind === "list") {
    const item = variable(level)
    for (const entry of node.items) unify(inferNode(entry, environment, level, constructorEnvironment), item, entry)
    type = list(item)
  } else if (node.kind === "tuple") type = tuple(node.items.map(item => inferNode(item, environment, level, constructorEnvironment)))
  else if (node.kind === "sequence") {
    const first = inferNode(node.first, environment, level, constructorEnvironment)
    unify(first, unit, node.first)
    type = inferNode(node.second, environment, level, constructorEnvironment)
  } else if (node.kind === "if") {
    const condition = inferNode(node.condition, environment, level, constructorEnvironment)
    const consequent = inferNode(node.consequent, environment, level, constructorEnvironment)
    const alternate = inferNode(node.alternate, environment, level, constructorEnvironment)
    unify(condition, bool, node.condition)
    unify(consequent, alternate, node)
    type = consequent
  } else if (node.kind === "function") {
    const parameter = variable(level + 1)
    const nested = new Map(environment)
    bindPattern(node.parameter, parameter, nested, level + 1)
    const result = inferNode(node.body, nested, level + 1, constructorEnvironment)
    type = arrow(parameter, result)
  } else if (node.kind === "apply") {
    const callee = inferNode(node.callee, environment, level, constructorEnvironment)
    const argument = inferNode(node.argument, environment, level, constructorEnvironment)
    const result = variable(level)
    unify(callee, arrow(argument, result), node.callee)
    type = result
  } else if (node.kind === "let") {
    const nested = new Map(environment)
    let value
    if (node.recursive) {
      const provisional = variable(level + 1)
      nested.set(node.name, { variables: [], type: provisional })
      value = inferNode(node.value, nested, level + 1, constructorEnvironment)
      unify(provisional, value, node.value)
    } else value = inferNode(node.value, environment, level + 1, constructorEnvironment)
    nested.set(node.name, generalize(value, level))
    type = inferNode(node.body, nested, level, constructorEnvironment)
  } else if (node.kind === "match") {
    const target = inferNode(node.target, environment, level, constructorEnvironment)
    const result = variable(level)
    for (const entry of node.cases) {
      const nested = new Map(environment)
      bindPattern(entry.pattern, target, nested, level)
      unify(inferNode(entry.body, nested, level, constructorEnvironment), result, entry.body)
    }
    type = result
  } else throw new LambError(`unsupported expression ${node.kind}`, node.start, node.end, "E_INTERNAL")
  node.type = prune(type)
  return node.type
}

export function infer(ast) {
  resetTypes()
  inferNode(ast, new Map(), 0, constructors(0))
  return ast
}

export { formatType }

File: ocaml/infer.ml
open Syntax
open Typesystem

type annotation = {
  location : span;
  datatype : datatype;
}

type result = {
  root : expression;
  datatype : datatype;
  annotations : annotation list;
}

exception Error of string * span

type environment = (string * scheme) list

let integer = primitive "int"
let floating = primitive "float"
let boolean = primitive "bool"
let stringtype = primitive "string"
let character = primitive "char"
let unittype = primitive "unit"

let lookup environment name location =
  match List.assoc_opt name environment with
  | Some scheme -> scheme
  | None -> raise (Error ("unbound value " ^ name, location))

let replace environment name scheme =
  (name, scheme) :: List.remove_assoc name environment

let constructors level =
  let item = variable level in
  [ "None", { quantified =
                begin match item with Variable value -> [value] | _ -> [] end;
              body = option item };
    "Some", { quantified =
                begin match item with Variable value -> [value] | _ -> [] end;
              body = arrow item (option item) } ]

let with_type_error location functionvalue =
  try functionvalue () with
  | Typesystem.Error message -> raise (Error (message, location))

let rec bind_pattern environment names value expected level =
  let bind pattern expected =
    bind_pattern environment names pattern expected level
  in
  match value.pattern with
  | Pwildcard -> environment, names
  | Pvariable name ->
      if List.mem name names then
        raise
          (Error
             ("variable " ^ name ^ " is bound several times in this pattern",
              value.pspan));
      replace environment name (monomorphic expected), name :: names
  | Pliteral literal ->
      let datatype =
        match literal with
        | Integer _ -> integer
        | Float _ -> floating
        | Boolean _ -> boolean
        | String _ -> stringtype
        | Character _ -> character
        | Unit -> unittype
      in
      with_type_error value.pspan (fun () -> unify expected datatype);
      environment, names
  | Pnil ->
      with_type_error value.pspan (fun () -> unify expected (list (variable level)));
      environment, names
  | Pcons (head, tail) ->
      let item = variable level in
      with_type_error value.pspan (fun () -> unify expected (list item));
      let environment, names =
        bind_pattern environment names head item level
      in
      bind_pattern environment names tail (list item) level
  | Ptuple items ->
      let types = List.map (fun _ -> variable level) items in
      with_type_error value.pspan (fun () -> unify expected (tuple types));
      List.fold_left2
        (fun (environment, names) pattern datatype ->
          bind_pattern environment names pattern datatype level)
        (environment, names)
        items
        types
  | Pconstructor ("None", None) ->
      with_type_error value.pspan (fun () -> unify expected (option (variable level)));
      environment, names
  | Pconstructor ("None", Some _) ->
      raise (Error ("constructor None expects no argument", value.pspan))
  | Pconstructor ("Some", None) ->
      raise (Error ("constructor Some expects one argument", value.pspan))
  | Pconstructor ("Some", Some argument) ->
      let item = variable level in
      with_type_error value.pspan (fun () -> unify expected (option item));
      bind argument item
  | Pconstructor (name, _) ->
      raise (Error ("unknown constructor " ^ name, value.pspan))

let infer root =
  reset ();
  let annotations = ref [] in
  let constructorenvironment = constructors 0 in
  let annotate node datatype =
    let datatype = prune datatype in
    annotations := { location = node.span; datatype } :: !annotations;
    datatype
  in
  let rec descend environment level node =
    let inferred =
      match node.expression with
      | Literal literal ->
          begin
            match literal with
            | Integer _ -> integer
            | Float _ -> floating
            | Boolean _ -> boolean
            | String _ -> stringtype
            | Character _ -> character
            | Unit -> unittype
          end
      | Variable name ->
          instantiate level (lookup environment name node.span)
      | Constructor name ->
          begin
            match List.assoc_opt name constructorenvironment with
            | Some scheme -> instantiate level scheme
            | None -> raise (Error ("unknown constructor " ^ name, node.span))
          end
      | Unary ("-", argument) ->
          let argumenttype = descend environment level argument in
          with_type_error argument.span (fun () -> unify argumenttype integer);
          integer
      | Unary (operator, _) ->
          raise (Error ("unknown unary operator " ^ operator, node.span))
      | Binary (operator, left, right) ->
          let lefttype = descend environment level left in
          let righttype = descend environment level right in
          begin
            match operator with
            | "+" | "-" | "*" | "/" | "mod" ->
                with_type_error left.span (fun () -> unify lefttype integer);
                with_type_error right.span (fun () -> unify righttype integer);
                integer
            | "^" ->
                with_type_error left.span (fun () -> unify lefttype stringtype);
                with_type_error right.span (fun () -> unify righttype stringtype);
                stringtype
            | "&&" | "||" ->
                with_type_error left.span (fun () -> unify lefttype boolean);
                with_type_error right.span (fun () -> unify righttype boolean);
                boolean
            | "=" | "<>" ->
                with_type_error node.span (fun () -> unify lefttype righttype);
                boolean
            | "<" | ">" | "<=" | ">=" ->
                with_type_error node.span (fun () -> unify lefttype righttype);
                if not (is_ordered lefttype) then
                  raise
                    (Error
                       ( "operator " ^ operator ^ " does not order "
                         ^ format lefttype,
                         node.span ));
                boolean
            | _ -> raise (Error ("unknown operator " ^ operator, node.span))
          end
      | Cons (head, tail) ->
          let headtype = descend environment level head in
          let tailtype = descend environment level tail in
          with_type_error tail.span (fun () -> unify tailtype (list headtype));
          tailtype
      | List items ->
          let itemtype = variable level in
          List.iter
            (fun item ->
              let datatype = descend environment level item in
              with_type_error item.span (fun () -> unify datatype itemtype))
            items;
          list itemtype
      | Tuple items ->
          tuple (List.map (descend environment level) items)
      | Sequence (first, second) ->
          let firsttype = descend environment level first in
          with_type_error first.span (fun () -> unify firsttype unittype);
          descend environment level second
      | Conditional (condition, consequent, alternate) ->
          let conditiontype = descend environment level condition in
          let consequenttype = descend environment level consequent in
          let alternatetype = descend environment level alternate in
          with_type_error condition.span (fun () -> unify conditiontype boolean);
          with_type_error node.span (fun () -> unify consequenttype alternatetype);
          consequenttype
      | Function (parameter, body) ->
          let parametertype = variable (level + 1) in
          let nested, _ =
            bind_pattern environment [] parameter parametertype (level + 1)
          in
          let resulttype = descend nested (level + 1) body in
          arrow parametertype resulttype
      | Apply (callee, argument) ->
          let calleetype = descend environment level callee in
          let argumenttype = descend environment level argument in
          let resulttype = variable level in
          with_type_error callee.span (fun () ->
              unify calleetype (arrow argumenttype resulttype));
          resulttype
      | Let (binding, body) ->
          let valuetype, nestedenvironment =
            if binding.recursive then
              let provisional = variable (level + 1) in
              let recursiveenvironment =
                replace environment binding.name (monomorphic provisional)
              in
              let valuetype =
                descend recursiveenvironment (level + 1) binding.value
              in
              if not (equal provisional valuetype) then
                with_type_error binding.value.span (fun () ->
                    unify provisional valuetype);
              valuetype, recursiveenvironment
            else
              descend environment (level + 1) binding.value, environment
          in
          let nested =
            replace nestedenvironment binding.name (generalize level valuetype)
          in
          descend nested level body
      | Match (target, cases) ->
          let targettype = descend environment level target in
          let resulttype = variable level in
          List.iter
            (fun branch ->
              let nested, _ =
                bind_pattern environment [] branch.casepattern targettype level
              in
              let branchtype = descend nested level branch.casebody in
              with_type_error branch.casebody.span (fun () ->
                  unify resulttype branchtype))
            cases;
          resulttype
    in
    annotate node inferred
  in
  let datatype = descend [] 0 root in
  { root; datatype = prune datatype; annotations = List.rev !annotations }

let type_at (result : result) location =
  result.annotations
  |> List.find_opt (fun annotation -> annotation.location = location)
  |> Option.map (fun (annotation : annotation) -> annotation.datatype)

let format_result (result : result) =
  format result.datatype
